### DISCUSSION TITLE
Redundant calculations of objective functions and constraints

### DISCUSSION THREAD

**Discussion by BenVanDerMeer** [coin-or/Ipopt#725] [Q&A] [answered]
I'm using automatic differentiation to calculate the derivative of my objective function and constraints. As such I effectively have four different functions:
1. `calculate_f`
2. `calculate_f_and_grad_f`
3. `calculate_g`
4. `calculate_g_and_jac_g`

Notice that the functions with automatic differentiation (2 and 4) calculate both the objective function/constraint and the derivatives. But when interfacing with Ipopt there's no way to tell whether I should also calculate the derivatives or just the objection/functions constraints. In my current setup I simply call `calculate_f` in `eval_f`, `calculate_f_and_grad_f` in `eval_grad_f` and likewise for the constraints. This results in an extra calculation of `f` as it is needs to be calculated twice for each step. I could instead perform `calculate_f_and_grad_f` in `eval_f` and save the gradient to be given to `eval_grad_f`, but this would sometimes result in the gradient of `f` being calculated when it wasn't necessary. The crux of the problem is that there's no way to tell if Ipopt is going to ask for the gradient of the same spot in `eval_f` or `eval_g`, which it always seem to call before `eval_grad_f` or `eval_jac_g`. Note that the `new_x` parameter doesn't help me here. Any ideas of how to get around this issue? Thanks.

**Comment by svigerske** [comment, marked as answer]
A reason that Ipopt does not tell in advance whether it will also need a gradient is that it typically does not know itself. In my understanding, Ipopt tries several points during line search. To decide whether to accept a point as next iterate, it will need only function values (`eval_f` and `eval_g`). If a point is accepted, then it will need also the first and second derivatives to continue (i.e., calculate the next search direction).

So whether to always compute both function values and gradient/Jacobian with the risk that the latter is thrown away, or having to essentially recompute the function values when doing the work for gradient/Jacobian only when requested, depends on how much effort these steps are and how many trial points Ipopt needs to go through before accepting a point. Maybe you can implement both and switch to what seems better according to the instance and Ipopts behavior. Maybe you can find some pattern, like Ipopt typically not requesting the gradient/Jacobian for x new points after a point where it needed these (x being an estimate on the number of unsuccessful trials in line search).

**Reply by BenVanDerMeer** [reply]
I see, thanks Stefan. I had a feeling it would be something like that.
